applications/MappingApplication/custom_utilities/mapping_matrix_builder.h

```diff
@@ -0,0 +1,131 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Philipp Bucher, Jordi Cotela
+//
+// See Master-Thesis P.Bucher
+// "Development and Implementation of a Parallel
+//  Framework for Non-Matching Grid Mapping"
+
+#if !defined(KRATOS_MAPPING_MATRIX_BUILDER_H)
+#define  KRATOS_MAPPING_MATRIX_BUILDER_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/interface_vector_container.h"
+#include "custom_utilities/mapper_local_system.h"
+#include "custom_utilities/mapper_flags.h"
+
+namespace Kratos
+{
+///@addtogroup MappingApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+/// Object for constructing the mapping-system, equivalent to the BuilderAndSolver
+/** The mapping-system is constructed from the the MapperLocalSystems, this class
+ * and its derived objects take care of this.
+*/
+template<class TSparseSpace, class TDenseSpace>
+class MappingMatrixBuilder
+{
+    public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of MappingMatrixBuilder
+    KRATOS_CLASS_POINTER_DEFINITION(MappingMatrixBuilder);
+
+    typedef typename TSparseSpace::MatrixType TMappingMatrixType;
+    typedef typename TSparseSpace::VectorType TSystemVectorType;
+
+    typedef Kratos::unique_ptr<TMappingMatrixType> TMappingMatrixUniquePointerType;
+    typedef Kratos::unique_ptr<TSystemVectorType> TSystemVectorUniquePointerType;
+
+    typedef Kratos::unique_ptr<MapperLocalSystem> MapperLocalSystemPointer;
+    typedef std::vector<MapperLocalSystemPointer> MapperLocalSystemPointerVector;
+    typedef Kratos::shared_ptr<MapperLocalSystemPointerVector> MapperLocalSystemPointerVectorPointer;
+
+    typedef InterfaceVectorContainer<TSparseSpace, TDenseSpace> InterfaceVectorContainerType;
+    typedef Kratos::unique_ptr<InterfaceVectorContainerType> InterfaceVectorContainerPointerType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    MappingMatrixBuilder();
+
+    /// Destructor.
+    virtual ~MappingMatrixBuilder() = default;
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void BuildMappingMatrix(
```

Comment:
Did you consider making this a templated function instead of a class? It seems a bit odd to couple the builder to the object.